src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -932,37 +1077,50 @@ public virtual void Disconnect()
             using (Workspace?.Services.GetService<IGlobalOperationNotificationService>()?.Start("Disconnect Project"))
             {
                 // No sense in reloading any metadata references anymore.
-                foreach (var cancellationTokenSource in _changedReferencesPendingUpdate.Values)
+                foreach (var cancellationTokenSource in ChangedReferencesPendingUpdate.Values)
                 {
                     cancellationTokenSource.Cancel();
                 }
 
-                _changedReferencesPendingUpdate.Clear();
+                ChangedReferencesPendingUpdate.Clear();
 
                 var wasPushing = _pushingChangesToWorkspaceHosts;
 
                 // disable pushing down to workspaces, so we don't get redundant workspace document removed events
                 _pushingChangesToWorkspaceHosts = false;
 
+                ImmutableArray<IVisualStudioHostDocument> documents;
+                ImmutableArray<IVisualStudioHostDocument> additionalDocuments;
+                ImmutableArray<VisualStudioMetadataReference> metadataReferences;
+                ImmutableArray<VisualStudioAnalyzer> analyzers;
+
+                lock (_gate)
```

Comment:
I wasn't sure about the perf for the methods below, especially if the number of docs/references/analyzers is large, so avoided taking the lock for the whole thing. Probably doesn't matter as the project system is anyway disposing this project, I can switch to doing this entirely under the lock.